docs/modules/architecture/pages/compact-binary-specification.adoc

```diff
@@ -0,0 +1,719 @@
+|(number \| null)[] \| null
+|[]*int8
+|
+|Array of nullable int8s
+|no
+|nullable-int16[]
+|Short[] 
+|optional<std::vector<optional<int16_t>>>
+|short?[]?
+|typing.Optional[typing.List[typing.Optional[int]]]
+|(number \| null)[] \| null
+|[]*int16
+|
+|Array of nullable i1int6s
+|no
+|nullable-int32[]
+|Integer[] 
+|optional<std::vector<optional<int32_t>>>
+|int?[]?
+|typing.Optional[typing.List[typing.Optional[int]]]
+|(number \| null)[] \| null
+|[]*int32
+|
+|Array of nullable int32s
+|no
+|nullable-int64[]
+|Long[] 
+|optional<std::vector<optional<int64_t>>> 
+|long?[]?
+|typing.Optional[typing.List[typing.Optional[int]]]
+|(Long \| null)[] \| null
+|[]*int64
+|
+|Array of nullable int64s
+|no
+|nullable-float32[]
+|Float[] 
+|optional<std::vector<optional<float>>> 
+|float?[]?
+|typing.Optional[typing.List[typing.Optional[float]]]
+|(number \| null)[] \| null
+|[]*float32
+|
+|Array of nullable float32s
+|no
+|nullable-float64[]
+|Double[] 
+|optional<std::vector<optional<double>>> 
+|double?[]?
+|typing.Optional[typing.List[typing.Optional[float]]]
+|(number \| null)[] \| null
+|[]*float64
+|
+|Array of nullable float64
+|no
+|===
+
+
+=== Type Ids
+
+Each type supported in the wire format has its type id. The type ids are used while constructing the schemas, performing type checks for user access to fields, and are exposed to the user as a public API.
+
+To distinguish different types supported in the wire-level format, a new enum called `FieldKind` is introduced which has different ids for different types compared to the `FieldType` enum.
+
+The `FieldType` enum has different ways to represent ids for types and their arrays and has some unfortunate public methods which would not fit the new format. (for example `FieldType#DATE` returns `getTypeSize` wrong for the new format). It is desired to have a new enum that does not have those problems and is more open to type additions in the future.
+
+Note that, the new `FieldKind` enum with those new ids will be used in the `GenericRecord` API for both `Portable` and the new format. `FieldType` and `Portable` will be deprecated in the future.
+
+=== Unsigned Integers
+
+There is support for unsigned integer types. The representation of unsigned integers is done with the smallest signed integer type that can represent it (except for `u64`, since there is no such type). So, basically,
+
+* `u8` is represented by → `i16`
+* `u16` is represented by → `i32`
+* `u32` is represented by → `i64`
+* `u64` is represented by → `BigInteger`
+
+=== Enums
+
+There is no support for enums in the wire-level format. The representation of enums is left to the user. 
+
+=== Nullable Primitives
+
+There is support for nullable primitive types as they can be useful in SQL and might play nicely with languages that have a concept to represent them. For example, a C# user might use `Nullable<PrimitiveType>`, or Java users might use the class representation of primitive types.
```

Comment:
Should we talk about the languages in this reference? I thought it would be something language agnostic. I believe, instead of saying that we do it for this and that language, we should just specify how we do it